automation/src/test/scala/org/broadinstitute/dsde/test/api/SamApiSpec.scala

```diff
@@ -421,6 +419,94 @@ class SamApiSpec extends FreeSpec with BillingFixtures with Matchers with ScalaF
+        5.minutes, 5.seconds)
+
+      Sam.user.setPolicyMembers(authDomainId, "admin", Set(inAuthDomainUser.email, inBothUser.email))(inBothUserAuthToken)
+      awaitAssert(
+        Await.result(googleDirectoryDAO.listGroupMembers(authDomainAdminEmail.head), 5.minutes)
+          .getOrElse(Set.empty) should contain theSameElementsAs Set(inBothUserProxy.value, inAuthDomainUserProxy.value),
+        5.minutes, 5.seconds)
+
+      val resourceTypeName = "workspace"
+      val resourceId = UUID.randomUUID.toString
+      val ownerPolicyName = "owner"
+      val policies = Map(ownerPolicyName -> AccessPolicyMembership(Set(inBothUser.email), Set.empty, Set(ownerPolicyName)))
+      val resourceRequest = CreateResourceRequest(resourceId, policies, Set(authDomainId))
+
+      // Create constrained resource
+      Sam.user.createResource(resourceTypeName, resourceRequest)(inBothUserAuthToken)
+      register cleanUp Sam.user.deleteResource(resourceTypeName, resourceId)(inBothUserAuthToken)
+
+      Sam.user.addUserToResourcePolicy(resourceTypeName, resourceId, ownerPolicyName, inPolicyUser.email)(inBothUserAuthToken)
+      val resourcePolicies = Sam.user.listResourcePolicies(resourceTypeName, resourceId)(inBothUserAuthToken)
+      val resourceOwnerEmail = resourcePolicies.collect {
+        case SamModel.AccessPolicyResponseEntry(_, policy, email) if policy.memberEmails.nonEmpty => email
+      }
+      assert(resourceOwnerEmail.size == 1)
+      Sam.user.syncResourcePolicy(resourceTypeName, resourceId, ownerPolicyName)(inBothUserAuthToken)
+
+      // Google should only know about the user that is in both the auth domain group and the constrained policy
+      awaitAssert(
+        Await.result(googleDirectoryDAO.listGroupMembers(resourceOwnerEmail.head), 5.minutes)
+          .getOrElse(Set.empty) should contain theSameElementsAs Set(inBothUserProxy.value),
+        5.minutes, 5.seconds)
+    }
+
+    "should synchronize the all users group for public policies" in {
+      val resourceId = UUID.randomUUID.toString
+      val user1 = UserPool.chooseStudent
+      val user1AuthToken = user1.makeAuthToken()
+      val user1Proxy = Sam.user.proxyGroup(user1.email)(user1AuthToken)
+      val allUsersGroupEmail = Sam.user.getGroupEmail("All_Users")(user1AuthToken)
+
+      val resourceTypeName = "managed-group"
+      val adminPolicyName = "admin"
+      val adminNotifierPolicyName = "admin-notifier"
+
+      Sam.user.createGroup(resourceId)(user1AuthToken)
+      register cleanUp Sam.user.deleteGroup(resourceId)(user1AuthToken)
+
+      val policies = Sam.user.listResourcePolicies(resourceTypeName, resourceId)(user1AuthToken)
+      val adminPolicy = policies.filter(_.policyName equals adminPolicyName).last
+      val adminNotifierPolicy = policies.filter(_.policyName equals adminNotifierPolicyName).last
+
+      awaitAssert(
+        Await.result(googleDirectoryDAO.listGroupMembers(adminPolicy.email), 5.minutes)
```

Comment:
it would nice if you can make these operations in a for comprehension, and await only once in the end. It'll be a lot easier for refactoring later (both less `unsafeRunSync` or `unsafeToFuture` if we're going to use `AsyncFlatSpec`)